### PR TITLE
[wasm-split] Initial instrumentation

### PR DESCRIPTION
**Pull request by tlively**
Implement an instrumentation pass that records the timestamp at which each
defined function is first called. Timestamps are not actual time, but rather
snapshots of a monotonically increasing counter. The instrumentation exports a
function that the embedder can call to dump the profile data into a memory
buffer at a given offset and size. The function returns the total size of the
profile data so the embedder can know how much to read out of the buffer or how
much it needs to grow the buffer.

Parsing and using the profile is left as future work, as is recording a hash of
the input file that will be used to guard against accidentally instrumenting one
module and trying to use the resulting profile to split a different module.

Since GitHub doesn't support stacked diffs yet this PR includes content from #3359.
Please review only the commits beginning with the eponymous commit.

cc @sbc100 @awtcode

### CODE REVIEW

File: src/tools/wasm-split.cpp [hunk 1]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
   options.applyFeatures(wasm);
 }
 
+// Add a global monotonic counter and a timestamp global for each function, code
+// at the beginning of each function to set its timestamp, and a new exported
+// function for dumping the profile data.
+struct FuncInstrumentor : public Pass {
```
Comment (kripken) [inline src/tools/wasm-split.cpp:267]: This name seems too generic - how about `TimeInstrumentor`?

File: src/tools/wasm-split.cpp [hunk 2]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  counterGlobal = Names::getValidGlobalName(*wasm, "monotonic_counter");
+  functionGlobals.reserve(wasm->functions.size());
+  ModuleUtils::iterDefinedFunctions(*wasm, [&](Function* func) {
+    functionGlobals.push_back(Names::getValidGlobalName(
+      *wasm, std::string(func->name.c_str()) + "_timestamp"));
+  });
+
+  // Create and add new globals
+  auto addGlobal = [&](Name name) {
+    auto global = std::make_unique<Global>();
```
Comment (kripken) [inline src/tools/wasm-split.cpp:308]: please use `builder.makeGlobal`

File: src/tools/wasm-split.cpp [hunk 3]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  // Inject the following code at the beginning of each function to advance the
+  // monotonic counter and set the function's timestamp if it hasn't already
+  // been set.
+  //
+  //   (if (i32.eqz (global.get $timestamp))
+  //     (block
+  //       (global.set $monotonic_counter
+  //         (i32.add
+  //           (global.get $monotonic_counter)
+  //           (i32.const)
```
Comment (kripken) [inline src/tools/wasm-split.cpp:332]: missing `1`

File: src/tools/wasm-split.cpp [hunk 4]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  //     )
+  //   )
+  Builder builder(*wasm);
+  auto globalIt = functionGlobals.begin();
+  ModuleUtils::iterDefinedFunctions(*wasm, [&](Function* func) {
+    func->body = builder.makeSequence(
+      builder.makeIf(
+        builder.makeUnary(EqZInt32,
+                          builder.makeGlobalGet(*globalIt, Type::i32)),
+        builder.blockify(
```
Comment (kripken) [inline src/tools/wasm-split.cpp:347]: could be `makeSequence` (`blockify` does more things, which you do not need here)

File: src/tools/wasm-split.cpp [hunk 5]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+    offset += 4;
+  }
+
+  writeProfile->body = builder.blockify(
+    builder.makeIf(builder.makeBinary(GeUInt32, getSize(), profileSizeConst()),
+                   writeData),
+    profileSizeConst());
+
+  // Create an export for the function
+  auto ex = std::make_unique<Export>();
```
Comment (kripken) [inline src/tools/wasm-split.cpp:413]: please use `builder.makeExport`

File: src/tools/wasm-split.cpp [hunk 6]
```diff
@@ -245,8 +261,195 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  if (wasm.getExportOrNull(options.profileExport) != nullptr) {
+    Fatal() << "error: Export " << options.profileExport << " already exists.";
+  }
+
+  // TODO: calculate module hash.
+  uint64_t moduleHash = 0;
+  PassRunner runner(&wasm, options.passOptions);
+  runner.add(
+    std::make_unique<FuncInstrumentor>(options.profileExport, moduleHash));
+  runner.run();
```
Comment (kripken) [inline src/tools/wasm-split.cpp:446]: I think you can replace these 3 lines with `FuncInstrumentor(..).run(runner, module);`

**Review by kripken — APPROVED** [review]
lgtm % comments

File: src/tools/wasm-split.cpp (hunk 2 above)
Comment (tlively) [reply src/tools/wasm-split.cpp:308]: Sounds good, but I'll wait for #3391 to land first ;)

File: src/tools/wasm-split.cpp (hunk 1 above)
Comment (tlively) [reply src/tools/wasm-split.cpp:267]: `Func` is supposed to differentiate it from a hypothetical future `BlockInstrumenter`. But given that this is the only instrumentation in the tool right now, how about doubling down on the generality and changing the name to "Instrumenter"? This is entirely unambiguous in the context of wasm-split.

File: src/tools/wasm-split.cpp [hunk 7]
```diff
@@ -245,8 +261,189 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  this->wasm = wasm;
+  addGlobals();
+  instrumentFuncs();
+  addProfileExport();
+}
+
+void Instrumenter::addGlobals() {
+  // Create fresh global names
+  counterGlobal = Names::getValidGlobalName(*wasm, "monotonic_counter");
+  functionGlobals.reserve(wasm->functions.size());
```
Comment (aheejin) [inline src/tools/wasm-split.cpp:300]: Wouldn't this be the size for both imported and defined functions?

File: src/tools/wasm-split.cpp [hunk 8]
```diff
@@ -245,8 +261,189 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+  writeProfile->hasExplicitName = true;
+  writeProfile->setLocalName(0, "addr");
+  writeProfile->setLocalName(1, "size");
+
+  // Calculate the size of the profile:
+  //   8 bytes module hash +
+  //   4 bytes for the timestamp for each function
+  size_t numDefinedFunctions = 0;
+  ModuleUtils::iterDefinedFunctions(*wasm,
+                                    [&](Function*) { numDefinedFunctions++; });
```
Comment (aheejin) [inline src/tools/wasm-split.cpp:378]: Maybe it would be better to have a helper function that returns the number of imported / defined functions...

**Review by aheejin — APPROVED** [review]
Nice!

File: src/tools/wasm-split.cpp [hunk 9]
```diff
@@ -245,8 +261,189 @@ void parseInput(Module& wasm, const WasmSplitOptions& options) {
...
+    return builder.makeConst(int32_t(profileSize));
+  };
+
+  // Write the hash followed by all the time stamps
+  Expression* writeData =
+    builder.makeStore(8, 0, 1, getAddr(), hashConst(), Type::i64);
+
+  uint32_t offset = 8;
+  for (const auto& global : functionGlobals) {
+    writeData = builder.blockify(
```
Comment (aheejin) [inline src/tools/wasm-split.cpp:396]:
```suggestion
    writeData = builder.makeSequence(
```
I guess this can be `makeSequence` too?

Comment (tlively) [reply src/tools/wasm-split.cpp:396]: No, this one should be a `blockify`. The difference is that `makeSequence` always introduces a new block, but `blockify` appends to the first item if it is already a block.

File: src/tools/wasm-split.cpp (hunk 1 above)
Comment (kripken) [reply src/tools/wasm-split.cpp]: sgtm. I didn't realize the intended comparison to blocks.

File: src/tools/wasm-split.cpp (hunk 9 above)
Comment (aheejin) [reply src/tools/wasm-split.cpp:396]: Ah you're right; I thought `writeData` was always a `store`, but this is within a for loop.

File: src/tools/wasm-split.cpp (hunk 7 above)
Comment (tlively) [reply src/tools/wasm-split.cpp:300]: Yes, but it's probably okay to over-allocate a bit here.

File: src/tools/wasm-split.cpp (hunk 8 above)
Comment (tlively) [reply src/tools/wasm-split.cpp]: I changed this to just use `functionGlobals.size()`.

**Pull request merged**